std/hash/wyhash.zig

```diff
@@ -0,0 +1,99 @@
+const std = @import("std");
+const mem = std.mem;
+
+const primes = [_]u64{
+    0xa0761d6478bd642f,
+    0xe7037ed1a0b428db,
+    0x8ebc6af09c88c6e3,
+    0x589965cc75374cc3,
+    0x1d8e4e27c47d124f,
+};
+
+fn read_bytes(comptime bytes: u8, data: []const u8) u64 {
+    return mem.readVarInt(u64, data[0..bytes], @import("builtin").endian);
+}
+
+fn read_8bytes_swapped(data: []const u8) u64 {
+    return (read_bytes(4, data) << 32 | read_bytes(4, data[4..]));
+}
+
+fn mum(a: u64, b: u64) u64 {
+    var r: u128 = @intCast(u128, a) * @intCast(u128, b);
+    r = (r >> 64) ^ r;
+    return @truncate(u64, r);
+}
+
+fn mix0(a: u64, b: u64, seed: u64) u64 {
+    return mum(a ^ seed ^ primes[0], b ^ seed ^ primes[1]);
+}
+
+fn mix1(a: u64, b: u64, seed: u64) u64 {
+    return mum(a ^ seed ^ primes[2], b ^ seed ^ primes[3]);
+}
+
+pub fn hash(key: []const u8, initial_seed: u64) u64 {
+    var seed = initial_seed;
+
+    var i: usize = 0;
+    while (i + 32 <= key.len) : (i += 32) {
+        seed = mix0(
+            read_bytes(8, key[i..]),
+            read_bytes(8, key[i + 8 ..]),
+            seed,
+        ) ^ mix1(
+            read_bytes(8, key[i + 16 ..]),
+            read_bytes(8, key[i + 24 ..]),
+            seed,
+        );
+    }
+
+    const rem_len = @truncate(u5, key.len);
+    const rem_key = key[i..];
+    seed = switch (rem_len) {
```

Comment:
You could actually generate this code at compile-time pretty easily and I think it would be cool to change it at some point. Don't worry for the moment, though.

Reply:
That was my first approach but I missed something and got it wrong, then went for a more direct port.